TNL-4412 user is not enrolled as a beta tester when account is unactivated.

attiyaIshaque opened the pull request:
## [Batch Beta Tester Addition indicates success when inactive user added - TNL-4412](https://openedx.atlassian.net/browse/TNL-4412)

### Description
This PR fixes the when the inactive user is added in the *Beta tester addition*, it would show the message `These users could not be added as beta testers because their accounts are not yet activated:`

### How to Test?

**Stage** 
1. Go to test course: https://courses.edx.org/courses/course-v1:edX+Matt101+2016_T1/instructor#view-membership 
2. Enter any `inactive username` in the Batch Beta Tester Addition box.
3. Click Add beta testers - You'll see a success message
4. Under the Course Team Management area, select Beta Testers from the dropdown. 
5. Inactive users will not be listed 

**Sandbox**
https://beta-testers.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/instructor#view-membership
Inactive username : _Attiya_
Active username : _staff_

**Screenshots**

**Before Fix**
<img width="595" alt="screen shot 2017-04-20 at 6 21 11 pm" src="https://cloud.githubusercontent.com/assets/7627421/25233545/940fc4e0-25f8-11e7-84a1-12d949635f4a.png">

**After Fix**
<img width="851" alt="screen shot 2017-04-24 at 2 20 04 pm" src="https://cloud.githubusercontent.com/assets/7627421/25330452/2b3ced40-28f9-11e7-859c-fd20cb5e19d1.png">



### Tests
 - [x] Acceptance Test

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] @marcotuts  
- [x] @adampalay 
- [x] @noraiz-anwar 

FYI: @mduboseedx

### Post-review
- [x] Rebase and squash commits

attiyaIshaque commented:
jenkins run js

attiyaIshaque commented:
jenkins run bokchoy

attiyaIshaque commented:
jenkins run js

adampalay commented on lms/djangoapps/instructor/enrollment.py:
```diff
@@ -196,20 +196,14 @@ def send_beta_role_email(action, user, email_params):
         email_params['email_address'] = user.email
         email_params['full_name'] = user.profile.name
-
-    elif action == 'remove':
```
nice clean up!

attiyaIshaque replied:
Thanks.

adampalay commented on lms/static/js/instructor_dashboard/membership.js:
```diff
@@ -463,11 +463,26 @@ such that the value can be defined later than this assignment (file load order).
+                            inActive = true;
+                        }
+                    }
+                    if (inActive) {
```
instead of maintaining a separate `inActive` boolean, can you just check that `inActiveUsers` isn't empty? That simplifies the code a little

adampalay commented on lms/static/js/instructor_dashboard/membership.js:
```diff
@@ -463,11 +463,26 @@ such that the value can be defined later than this assignment (file load order).
                         sr = successes[j];
-                        results.push(sr.identifier);
+                        if (sr.is_active) {
+                            results.push(sr.identifier);
```
~should `results` be renamed to `activeUsers`?~ (see comment below)

adampalay commented on lms/static/js/instructor_dashboard/membership.js:
```diff
@@ -463,11 +463,26 @@ such that the value can be defined later than this assignment (file load order).
+                    for (k = 0; k < inActiveUsers.length; k++) {
+                        results.push(inActiveUsers[k]);
                     }
                     return results;
```
ah, I see why you were adding the active users to `results`. Disregard my last comment then :)

adampalay commented on lms/static/js/instructor_dashboard/membership.js:
```diff
@@ -463,11 +463,26 @@ such that the value can be defined later than this assignment (file load order).
+                            'testers because their accounts are not yet activated:';
+                        results.push(message);
+                    }
+                    for (k = 0; k < inActiveUsers.length; k++) {
```
Cleaner to use [`concat`](https://www.w3schools.com/jsref/jsref_concat_array.asp)

attiyaIshaque replied:
Really Awesome. 👍

adampalay commented on lms/djangoapps/instructor/enrollment.py:
```diff
@@ -196,20 +196,14 @@ def send_beta_role_email(action, user, email_params):
         raise ValueError("Unexpected action received '{}' - expected 'add' or 'remove'".format(action))
-
-    send_mail_to_student(user.email, email_params, language=get_user_email_language(user))
+    if not (not user.is_active and action == 'add'):
```
might be easier to understand as:
`user.is_active or action != 'add'`

Or
```python
trying_to_add_inactive_user = not user.is_active and action == 'add'
if not trying_to_add_inactive_user:
  ...
```

attiyaIshaque commented:
@adampalay I have addressed all the comments.

marcotuts commented:
Some small design level feedback from me. 

# We should render the sentence: "These users could not be added as beta testers" as the same heading level as "These users were successfully added..."
# If there are no students who are successfully adde that heading shouldn't render. In the image below, if I don't read the "These users could not be added." I might think they were added because of the larger heading. 
![image](https://cloud.githubusercontent.com/assets/2023680/25243187/344a43ec-25ca-11e7-95fb-823430a5c31b.png)

attiyaIshaque commented:
@marcotuts okay i will change it. Thanks for pointing this.

noraiz-anwar commented on common/test/acceptance/tests/lms/test_lms_instructor_dashboard.py:
```diff
@@ -213,6 +213,37 @@ def test_auto_enroll_csv_a11y(self):
         self.auto_enroll_section.a11y_audit.check_for_accessibility_errors()
 
 
+class BatchBetaTestersTest(BaseInstructorDashboardTest):
```
Not sure if it should be covered in this scope but there are other scenarios related to batch beta testers related to you change. For example test for when there are active and inactive users together to be added as beta testers.

attiyaIshaque replied:
Yeah, you are right. I will add one more test.

attiyaIshaque commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -880,6 +886,44 @@ def get_notification_text(self):
+        self.q(css=username_selector).fill(username)
+
+        # Verify enrollment button is present before clicking
+        EmptyPromise(
```
Yeah there is an other way, but here i am just checking if the element is preset before clicking and i am checking it by the `EmptyPromise `.

noraiz-anwar replied:
Why not with `wait_for_element_visibility`?

attiyaIshaque replied:
Yeah you are right, but it's just a another way.

adampalay replied:
I suggest using the built-in util, too

attiyaIshaque commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -880,6 +886,44 @@ def get_notification_text(self):
+    def is_browser_on_page(self):
+        return self.q(css=self.batch_beta_tester_heading_selector).present
+
+    def fill_beta_tester_text_box(self, username):
```
Actually we are already in the beta section on the membership page.

noraiz-anwar replied:
How do we know that?

attiyaIshaque replied:
Because of this `is_browser_on_page `. So what's your suggestion if we changes the name?

attiyaIshaque commented:
@noraiz-anwar I have addressed all your comments and added a new test. Please have a look.
@marcotuts I have updated the design and also updated the sandbox with new design.

noraiz-anwar commented on lms/static/js/instructor_dashboard/membership.js:
```diff
@@ -454,21 +456,40 @@ such that the value can be defined later than this assignment (file load order).
-                    $idsList.append($('<li/>', {
-                        text: identifier
-                    }));
+                    if (identifier === message) {
```
If we can do as described in comment below, that may make this block more easy to maintain and read.

attiyaIshaque replied:
Sure. I will add it.

attiyaIshaque commented on lms/static/js/instructor_dashboard/membership.js:
```diff
@@ -454,21 +456,40 @@ such that the value can be defined later than this assignment (file load order).
             };
             if (successes.length && dataFromServer.action === 'add') {
                 // Translators: A list of users appears after this sentence;
                 renderList(gettext('These users were successfully added as beta testers:'), (function() {
```
Yeah, firstly I consider it to separate the block `These users were successfully added as beta testers:`
But `renderList` method only render one list i.e either the `Active` or `Inactive`.
So that's why I have added the functionality of `Inactive users` under this part.

adampalay replied:
@attiyaIshaque , I think I agree with @noraiz-anwar  here; it's semantically a little weird to have the second list's header as an element of the first list. Can you make a `renderLists` method that will be able to combine two `renderList`s?

attiyaIshaque commented:
jenkins run js

attiyaIshaque commented:
jenkins run js

attiyaIshaque commented:
jenkins run js

attiyaIshaque commented:
jenkins run bokchoy

attiyaIshaque commented:
@adampalay , @marcotuts , @noraiz-anwar Please review again.

adampalay commented on lms/static/js/instructor_dashboard/membership.js:
```diff
@@ -463,13 +465,23 @@ such that the value can be defined later than this assignment (file load order).
+                        renderList(message, inActiveUsers);
+                    }
+                    if (activeUsers.length) {
+                        return activeUsers;
```
I feel like it would make sense to compile the list of `inactiveUsers` and `activeUsers` outside of the renderList function, and then just call `renderList` twice? Like:

```javascript
renderList(gettext('These users were successfully added as beta testers:'), activeUsers));
renderList(gettext('These users could not be added as beta testers because their accounts are not yet activated:'), inactiveUsers));
```

attiyaIshaque commented:
@adampalay @noraiz-anwar @marcotuts ping

adampalay commented on lms/static/js/instructor_dashboard/membership.js:
```diff
@@ -461,16 +461,26 @@ such that the value can be defined later than this assignment (file load order).
+                }
+                if (inActiveUsers.length) {
+                    // Translators: A list of users appears after this sentence;
+                    renderList(gettext('These users could not be added as beta testers ' +
```
I _think_ (but am not positive), that `gettext` pulls a static string, so I'm not sure you can pass a concatenated string like this to it.

adampalay commented:
@attiyaIshaque , couple comments from me

adampalay commented:
once tests and stuff pass 👍

attiyaIshaque commented:
@noraiz-anwar Ping.

attiyaIshaque commented:
jenkins run js
